Add notifications for when friends go online or offline

smoogipoo opened the pull request:
Supersedes https://github.com/ppy/osu/pull/31389
Requires https://github.com/ppy/osu-server-spectator/pull/255 for proper operation, but is backwards-compatible.

|         | Old Server           | New Server  |
| ------------- |:-------------:| :-----:|
| Old Client      | :green_circle:  | :green_circle: |
| New Client      | :green_circle:  | :green_circle: |

Tried to match the feel and operation of the same notification in osu!stable:

https://github.com/user-attachments/assets/20a63d0a-7897-4d26-8087-91946cad5d27

smoogipoo committed:
Display notification on friend presence changes

smoogipoo committed:
Add config option

peppy reviewed (commented):
only tested behaviour, not code yet. may require some re-thinking so going to leave it here for now.

smoogipoo commented on osu.Game/Online/FriendPresenceNotifier.cs:
```diff
@@ -0,0 +1,167 @@
+
+                        break;
+
+                    case NotifyDictionaryChangedAction.Remove:
```
Are you intending to receive a notification immediately when adding a friend? Is that the "predicament"? If so, then that's not something I can resolve, because the server will not know of changes to a user's friends list anyway until they reconnect.

As for the dashboard, I suppose the reason that's done is to save on memory which I'm not sure is super important. In any case it may be enough to only clear non-friend states.

smoogipoo replied:
Have applied a prospective fix for the dashboard.

peppy replied:
The issue is only the dashboard part, correct. Showing the notification when adding a friend is optimal if it can remain that way.

smoogipoo committed:
Keep friend states when stopping watching global activity

smoogipoo committed:
Add separate path for friend presence notifications

It proved to be too difficult to deal with the flow that clears user
states on stopping the watching of global presence updates. It's not
helped in the least that friends are updated via the API, so there's a
third flow to consider (and the timings therein - both server-spectator
and friends are updated concurrently).

Simplest is to separate the friends flow, though this does mean some
logic and state duplication.

peppy commented:
Is this series of PRs ready for re-review yet? Or are you still working on things?

smoogipoo commented:
It's ready for re-review.

peppy commented:
Testing looks solid this time around 👍 .

smoogipoo committed:
Simplify operations on local list

smoogipoo committed:
Remove unused using

smoogipoo commented on osu.Game/Online/API/APIAccess.cs:
```diff
@@ -610,8 +612,14 @@ public void UpdateLocalFriends()
+                // Remove non-friends from local lists.
+                friendsSet.Clear();
+                friendsSet.AddRange(res.Select(f => f.TargetID));
+                friends.RemoveAll(f => !friendsSet.Contains(f.TargetID));
```
Ah yeah. Not sure why I did this.
